### PR TITLE
Improve pinned messages in Labs

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes https://github.com/vector-im/element-web/issues/5405
Fixes https://github.com/vector-im/element-web/issues/5958
Fixes https://github.com/vector-im/element-web/issues/5871
Fixes https://github.com/vector-im/element-web/issues/17213
Fixes https://github.com/vector-im/element-web/issues/6926

https://user-images.githubusercontent.com/2403652/119676658-eafe2380-be35-11eb-9438-677a40fcaf2f.mov

image

Skipping design/product as it isn't leaving Labs yet

### CODE REVIEW

File: res/css/views/right_panel/_PinnedMessagesCard.scss [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2017 Travis Ralston
+Copyright 2021 The Matrix.org Foundation C.I.C.
```
Comment (turt2live) [inline res/css/views/right_panel/_PinnedMessagesCard.scss:2]: goodbye, self. 

(I'm okay with this happening - I don't want any of the old CSS associated with my name)

File: src/components/structures/NotificationPanel.tsx [hunk 2]
```diff
@@ -16,29 +14,25 @@ See the License for the specific language governing permissions and
...
 /*
  * Component which shows the global notification list using a TimelinePanel
  */
 @replaceableComponent("structures.NotificationPanel")
-class NotificationPanel extends React.Component {
-    static propTypes = {
-        onClose: PropTypes.func.isRequired,
-    };
-
+class NotificationPanel extends React.Component<IProps> {
```
Comment (turt2live) [inline src/components/structures/NotificationPanel.tsx:34]: we should move the `export default` up here, and probably use `PureComponent` for this

File: src/components/structures/RightPanel.tsx [hunk 3]
```diff
@@ -296,7 +295,13 @@ export default class RightPanel extends React.Component {
...
 
             case RightPanelPhases.NotificationPanel:
-                panel = <NotificationPanel onClose={this.onClose} />;
+                if (SettingsStore.getValue("feature_pinning")) {
+                    panel = <NotificationPanel onClose={this.onClose} />;
+                }
+                break;
+
+            case RightPanelPhases.PinnedMessages:
+                panel = <PinnedMessagesCard room={this.props.room} onClose={this.onClose} />;
```
Comment (turt2live) [inline src/components/structures/RightPanel.tsx:304]: I'm failing to understand what this change is really doing: if the labs flag is *disabled*, we refuse to show the notifications panel now? These two panels seem unrelated in nature.

File: src/components/views/right_panel/RoomHeaderButtons.tsx [hunk 4]
```diff
@@ -38,9 +42,35 @@ const ROOM_INFO_PHASES = [
     RightPanelPhases.Room3pidMemberInfo,
 ];
 
+const PinnedMessagesHeaderButton = ({ room, isHighlighted, onClick }) => {
+    const pinningEnabled = useSettingValue("feature_pinning");
+    const pinnedEvents = usePinnedEvents(pinningEnabled && room);
+    const readPinnedEvents = useReadPinnedEvents(pinningEnabled && room);
+    if (!pinningEnabled) return null;
```
Comment (turt2live) [inline src/components/views/right_panel/RoomHeaderButtons.tsx:49]: we can probably short circuit this by not reading pinned events if the setting isn't enabled. ie: move new lines 47 and 48 to below this condition:
```suggestion
    const pinningEnabled = useSettingValue("feature_pinning");
    if (!pinningEnabled) return null;

    const pinnedEvents = usePinnedEvents(pinningEnabled && room);
    const readPinnedEvents = useReadPinnedEvents(pinningEnabled && room);
```

File: src/components/views/rooms/PinnedEventTile.tsx [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    room: Room;
+    event: MatrixEvent;
+    onUnpinClicked?(): void;
+}
+
+const AVATAR_SIZE = 24;
+
+@replaceableComponent("views.rooms.PinnedEventTile")
+export default class PinnedEventTile extends React.Component<IProps> {
+    static contextType = MatrixClientContext;
```
Comment (turt2live) [inline src/components/views/rooms/PinnedEventTile.tsx:43]: explicit member access:
```suggestion
    public static contextType = MatrixClientContext;
```

**Review by turt2live — COMMENTED** [review]
overall seems fine, just a few comments relating to code style and a behavioural thing.

There's also a few mixed cases of `{words}` and `{ words }`, which I believe is a lint rule applied after this PR was introduced: would be good to normalize before landing.

File: res/css/views/right_panel/_PinnedMessagesCard.scss (hunk 1 above)
Comment (t3chguy) [reply res/css/views/right_panel/_PinnedMessagesCard.scss:2]: I mean, there is no old CSS here. All of it is brand new, I probably renamed the file to save me having to copy-paste the license :)

File: src/components/structures/RightPanel.tsx (hunk 3 above)
Comment (t3chguy) [reply src/components/structures/RightPanel.tsx:304]: Yeah I put the if around the wrong thing :D

File: src/components/views/right_panel/RoomHeaderButtons.tsx (hunk 4 above)
Comment (t3chguy) [reply src/components/views/right_panel/RoomHeaderButtons.tsx:49]: Hooks cannot be conditional.
But the two hooks here are effectively no-opd if you pass them a falsey as we are here.

The alternative would be tracking feature_pinning in RoomHeaderButtons but that's more lines of code due to the class setup than a `useSettingValue` hook call

Comment (turt2live) [reply src/components/views/right_panel/RoomHeaderButtons.tsx:49]: well that's not very fun :( 

This is fine, then.

**Review by turt2live — APPROVED** [review]

**Pull request merged**